### PR TITLE
Show shortened SHA1 in user interface hyperlink

### PR DESCRIPTION
**Pull request by MarkEWaite**
Attempt to keep the code DRY.

Please consider this mostly a solicitation for feedback, rather than a final submission that is ready to merge into the master branch.

Is this what you were envisioning by using Hamcrest instead of the style used in the existing test?

---
This change is [Reviewable](https://reviewable.io/reviews/jenkinsci/github-plugin/128)

### CODE REVIEW

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java [hunk 1]
```diff
@@ -42,7 +42,9 @@ void annotate(final GithubUrl url, final MarkupText text, final Entry change) {
 
         if (change instanceof GitChangeSet) {
             GitChangeSet cs = (GitChangeSet) change;
-            text.wrapBy("", " (<a href='" + url.commitId(cs.getId()) + "'>commit: " + cs.getId() + "</a>)");
+            final String id = cs.getId();
+            text.wrapBy("", " (<a href='" + url.commitId(id)
+                        + "'>commit: " + id.substring(0, Math.min(id.length(), 7)) + "</a>)");
```
Comment (KostyaSha) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java]: @lanwen maybe string format?

Comment (lanwen) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java]: yep!

**Comment by lanwen** [comment]
please add string formatting and i'll merge it

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java [hunk 2]
```diff
@@ -43,8 +43,14 @@ void annotate(final GithubUrl url, final MarkupText text, final Entry change) {
...
-            text.wrapBy("", " (<a href='" + url.commitId(id)
-                        + "'>commit: " + id.substring(0, Math.min(id.length(), 7)) + "</a>)");
+            final String innerText = "commit: " + id.substring(0, Math.min(id.length(), 7));
+            final String startHREF = "<a href='" + url.commitId(id) + "'>";
+            final String endHREF = "</a>":
+            text.wrapBy("", " ("
+                        + startHREF
+                        + innerText
+                        + endHREF
+                        + ")");
```
Comment (MarkEWaite) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java]: @lanwen was this the type of formatting you were envisioning? It seems to clarify the subcomponents of the markup, but I'm not sure that is what you were expecting?

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java [hunk 3]
```diff
@@ -43,8 +45,8 @@ void annotate(final GithubUrl url, final MarkupText text, final Entry change) {
         if (change instanceof GitChangeSet) {
             GitChangeSet cs = (GitChangeSet) change;
             final String id = cs.getId();
-            text.wrapBy("", " (<a href='" + url.commitId(id)
-                        + "'>commit: " + id.substring(0, Math.min(id.length(), 7)) + "</a>)");
+            final String innerText = "commit: " + id.substring(0, Math.min(id.length(), 7));
+            text.wrapBy("", format(" (<a href='%s'>%s</a>)", url.commitId(id), innerText));
```
Comment (MarkEWaite) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java]: Is this closer to the style you wanted in the code?

Comment (KostyaSha) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java]: No, String.format() that IDEA would suggest to replace automatically.

Comment (MarkEWaite) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java]: Unfortunately, I'm not an IDEA user.  Can you provide your preferred change to that line?

File: src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java [hunk 4]
```diff
@@ -42,7 +44,9 @@ void annotate(final GithubUrl url, final MarkupText text, final Entry change) {
 
         if (change instanceof GitChangeSet) {
             GitChangeSet cs = (GitChangeSet) change;
-            text.wrapBy("", " (<a href='" + url.commitId(cs.getId()) + "'>commit: " + cs.getId() + "</a>)");
+            final String id = cs.getId();
+            final String innerText = "commit: " + id.substring(0, Math.min(id.length(), 7));
```
Comment (lanwen) [inline src/main/java/com/coravy/hudson/plugins/github/GithubLinkAnnotator.java]: just inline it to string pattern to avoid string concatenation with "+"

**Comment by lanwen** [comment]
Perfectly!

**Pull request merged**